implement PauliStringTS and speed up OperatorTS2D

lukas-weber opened the pull request:
Adds PauliStringTS, which represents a translation-symmetric sum of a Pauli string in N-dimensions.

These sums are constructed by a much more efficient bit-shift algorithm (credit: ChatGPT), which knows the periods of the system at compile time. The improvements are applied to OperatorTS2D.

For calculating the 4th moment of a 5x5 Heisenberg model, I saw a 4× speedup.

It turns out that calculating the cardinality of the orbit of a string is nontrivial for 2D and higher, so I left it out. Instead, `opnorm` is defined using `trace_product(o,o)`, which seems to give the right result.

The changes should be completely API compatible, although I suggest `extent` and `shift(op, r1, r2, L1)` should be deprecated in favor of the n-dimensional `qubitsize` and `shift(op, Ls, shifts)`.

A natural next step is to refactor `OperatorTS1D` and `OperatorTS2D` into an N-dimensional equivalent. I would suggest to call it `PeriodicOperator` and `PeriodicPauliString` instead of `TS`. Happy to do that in case there is interest.

lkdvos commented on src/moments.jl:
```diff
@@ -73,32 +73,30 @@ end
                 for j in 0:L2-1
-                    p3 = rotate_lower(p2, i, j, L1)
-                    p, k = prod(p1, p3)
+                    shifted = shift(rep2, Val((L1, L2)), (i, j))
```
Doesn't this lead to a dynamic dispatch, kind of defeating the purpose of using the `Val`s?
Could you check to see if replacing this with regular integers actually makes things slower?

lkdvos commented on src/translation_symmetry.jl:
```diff
@@ -0,0 +1,119 @@
+@inline function shift(x::Unsigned, ::Val{Ls}, distances) where {Ls}
+    stride = 1
+    for (distance, L) in zip(distances, Ls)
+        x = shift(x, stride * distance, stride * L)
```
I think with this construction we lose the ability to const-propagate the `Ls` anyways, and it won't even reach the actual implementation of the `shift ` function.

lkdvos commented on src/translation_symmetry.jl:
```diff
@@ -0,0 +1,119 @@
+    rshift = stride - s
+
+    # masks for the region that crosses the periodic boundary after the shift
+    magicmask = (~(~zero(x) << s)) * ((~zero(x) ÷ ((one(x) << stride) - one(x))) >> rightzeros << rshift)
```
As a counter-proposal instead of the `Val`s: how about we precompute a (static?) array of these magicmasks and simply obtain them from memory? (that is, only in the case that we can actually see this line in a profiler)

lkdvos reviewed (commented):
This looks very nice to me!
I left a few comments and questions throughout, but additionally:

- Could we maybe try and add the case you are mentioning as a benchmark, so we can clearly monitor the performance? I'm happy to help set that up if you have a script somewhere.
- I am not a user so for me this is less important, but I would definitely agree with deprecating some of these old functions.

In particular, I would have guessed that you would simply have:

```julia
const OperatorTS1D = Operator{P<:PauliString1D,...}
```
etc...

In any case, I think the use of `Val` in its current form might be doing more harm than good, but this is just my intuition.

lukas-weber commented:
I added a benchmark, and merged most of the dimension specific code into one. Docs have been updated and converted to doctests. The old API should still work through some placeholder functions.

Regarding the `Val`s: I removed them and get the same performance, but only if everything is inlined. This is the case in Julia 1.11, in exactly the way it is written now. If this turns out to be too brittle, we can try doing something else. What counts is just that Ls is known at compile time, which allows circumventing all `div`s. The shifts seem to be negligible in terms of perfomance.

I also tried precomputing and storing the masks, either on the heap or an SVector, but the speed of that is approximately the same (a bit slower) than the current version. From this I guess that it’s all memory bound now.

I also applied the `trace_product` trick to the translation symmetric case.

nicolasloizeau commented:
This looks awesome thank you so much !
Not sure what's going on with the doctests but it seems very minor, we probably can remove the failing doctests ?

lukas-weber commented:
The failing test is comparing insignificant digits in the float output. I put a filter there that is supposed to filter them, but I think I didn’t do it correctly. I’ll look into it tomorrow.
